### PR TITLE
Dynamic Node Labels

### LINKED ISSUE #352: Distinguish between Node Labels and Node Features
**Description**
Currently both node labels and node features are dynamic node events, we want to distinguish them

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
Close #352

### CODE REVIEW

File: tgm/data/dg_data.py [hunk 1]
```diff
@@ -35,19 +35,23 @@ class DGData:
 
     Attributes:
         time_delta (TimeDeltaDG | str): Time granularity of the graph.
-        time (Tensor): 1D tensor of all event timestamps [num_edge_events + num_node_events].
+        time (Tensor): 1D tensor of all event timestamps [num_edge_events + num_node_events + num_node_labels].
         edge_mask (Tensor): Indices of edge events within `time`.
         edge_index (Tensor): Edge connections [num_edge_events, 2].
         edge_x (Tensor | None): Optional edge features [num_edge_events, D_edge].
-        node_mask (Tensor | None): Indices of node events within `time`.
+        node_x_mask (Tensor | None): Indices of node events within `time`.
```
Comment (ntgbaoo) [inline tgm/data/dg_data.py]:
```suggestion
        node_x_mask (Tensor | None): Mask of node events within `time`.
```

File: tgm/data/dg_data.py [hunk 2]
```diff
@@ -35,19 +35,23 @@ class DGData:
...
         node_x_nids (Tensor | None): Node IDs corresponding to node events [num_node_events].
         node_x (Tensor | None): Node features over time [num_node_events, D_node_dynamic].
+        node_y_mask (Tensor | None): Indices of node labels within `time`.
```
Comment (ntgbaoo) [inline tgm/data/dg_data.py]:
```suggestion
        node_y_mask (Tensor | None): Mask of node events within `time`.
```

File: tgm/data/dg_data.py (hunk 1 above)
Comment (benjaminnNgo) [reply tgm/data/dg_data.py]: I think `dynamic node features` is more informative? To me, a node label can also be a part of node events. Maybe I misunderstood the terminology

File: tgm/data/dg_data.py [hunk 3]
```diff
@@ -35,19 +35,23 @@ class DGData:
...
+        node_y_nids (Tensor | None): Node IDs corresponding to node labels [num_node_labels].
+        node_y (Tensor | None): Node features over time [num_node_labels, D_node_dynamic].
```
Comment (ntgbaoo) [inline tgm/data/dg_data.py]: To differentiate from `node_x`
```suggestion
        node_y (Tensor | None): Node labels over time [num_node_labels, D_node_dynamic].
```

File: tgm/data/dg_data.py [hunk 4]
```diff
@@ -270,10 +336,13 @@ def _maybe_cast_integral_tensor(x: Tensor, name: str) -> Tensor:
...
         # Ensure timestamps match number of global events
-        if self.time.ndim != 1 or self.time.shape[0] != num_edges + num_node_events:
+        if (
+            self.time.ndim != 1
+            or self.time.shape[0] != num_edges + num_node_events + num_node_labels
+        ):
             raise ValueError(
-                'timestamps must have shape [num_edges + num_node_events], '
-                f'got {num_edges} edges, {num_node_events} node_events, shape: {self.time.shape}. '
+                'timestamps must have shape [num_edges + num_node_events + num_node_labels], '
```
Comment (ntgbaoo) [inline tgm/data/dg_data.py]:
```suggestion
                'time must have shape [num_edges + num_node_events + num_node_labels], '
```

File: tgm/data/dg_data.py [hunk 5]
```diff
@@ -532,24 +634,37 @@ def from_raw(
         if node_x_time is not None:
             timestamps = torch.cat([timestamps, node_x_time])
             event_types = torch.cat([event_types, torch.ones_like(node_x_time)])
+        if node_y_time is not None:
+            timestamps = torch.cat([timestamps, node_y_time])
+            event_types = torch.cat(
+                [event_types, torch.full_like(node_y_time, fill_value=2)]
```
Comment (ntgbaoo) [inline tgm/data/dg_data.py:640]: Should we create constants somewhere to keep track of indicator values for different events?
`1` for dynamic node features events
`2` for node labels events

File: tgm/core/batch.py
Comment (ntgbaoo) [inline tgm/core/batch.py:1]: We should be consistent in the terminology here. In this DGBatch, we call `node_x` is `Dynamic node features for nodes` (line 25)

But in `DGData`, we call it `node event.`

File: tgm/data/dg_data.py (hunk 1 above)
Comment (ntgbaoo) [reply tgm/data/dg_data.py]: This is called `Dynamic node features for nodes`  in DGBatch.

File: tgm/data/dg_data.py (hunk 2 above)
Comment (ntgbaoo) [reply tgm/data/dg_data.py]: This is called `Dynamic node features for nodes`  in DGBatch.

File: docs/tutorials/dgraph_tutorial.md [hunk 6]
```diff
@@ -54,19 +54,27 @@ class DGData:
         edge_mask (Tensor): Mask of edge events within `time`.
         edge_index (Tensor): Edge connections [num_edge_events, 2].
         edge_x (Tensor | None): Optional edge features [num_edge_events, D_edge].
-        node_mask (Tensor | None): Mask of node events within `time`.
+        node_x_mask (Tensor | None): Indices of node events within `time`.
         node_x_nids (Tensor | None): Node IDs corresponding to node events [num_node_events].
         node_x (Tensor | None): Node features over time [num_node_events, D_node_dynamic].
+        node_y_mask (Tensor | None): Indices of node labels within `time`.
```
Comment (ntgbaoo) [inline docs/tutorials/dgraph_tutorial.md]:
```suggestion
        node_x_mask (Tensor | None): Mask of node events within `time`.
```

File: docs/tutorials/dgraph_tutorial.md [hunk 7]
```diff
@@ -54,19 +54,27 @@ class DGData:
         edge_mask (Tensor): Mask of edge events within `time`.
         edge_index (Tensor): Edge connections [num_edge_events, 2].
         edge_x (Tensor | None): Optional edge features [num_edge_events, D_edge].
-        node_mask (Tensor | None): Mask of node events within `time`.
+        node_x_mask (Tensor | None): Indices of node events within `time`.
```
Comment (ntgbaoo) [inline docs/tutorials/dgraph_tutorial.md]:
```suggestion
        node_x_mask (Tensor | None): Mask of node events within `time`.
```

Comment (ntgbaoo) [reply docs/tutorials/dgraph_tutorial.md]: Again, we need to be consistent whether to call `node_x` a node event or dynamic node features (we call it differently in `DGBatch` and `DGData`). I think dynamic node features is better, since node labels are also parts of node events?

**Review by benjaminnNgo — CHANGES_REQUESTED** [review]
`batch.dst` [here](https://github.com/tgm-team/tgm/blob/8559d68d1f1ebc1da47bbc750a8b1ccd398c6acf/docs/tutorials/hook_tutorial.md?plain=1#L51) need to be updated to `batch.edge_dst`

File: tgm/data/dg_data.py (hunk 5 above)
Comment (Jacob-Chmura) [reply tgm/data/dg_data.py:640]: They are not used anywhere except within this function, so I don't think its necessary. They are only used to create the masks.

File: docs/tutorials/dgraph_tutorial.md (hunk 7 above)
Comment (Jacob-Chmura) [reply docs/tutorials/dgraph_tutorial.md]: Agreed, changed the description convention to Dynamic Node Features

**Review by benjaminnNgo — APPROVED** [review]
The APIs now look very nice and clear to me. Thanks @Jacob-Chmura for the changes
